File: Temperature.java
// Time Complexity : O(N +k)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes

import java.util.Arrays;
import java.util.Stack;

public class Temperature {
    public int[] dailyTemperatures(int[] temperatures) {

      Stack<int[]> stack = new Stack<int[]>();
      int[] arr = new int[temperatures.length];
      Arrays.fill(arr,0);
      stack.push(new int[]{ temperatures[0], 0});

      for(int i=1; i< temperatures.length; i++){

        while(!stack.isEmpty() && stack.peek()[0] <  temperatures[i]){
          arr[stack.peek()[1]]= i-stack.peek()[1];
          stack.pop();
        }
        stack.push(new int[]{ temperatures[i], i});
      }

      return arr;
    }
  }

File: Temperature.py
#Monotonic stack solution
#Time complexity: O(2n) #For eg, 76,75,74,73,77
#Space complexity:O(n)
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        mono_stack=[]
        output=[0 for i in range(len(temperatures))]
        for i in range(0, len(temperatures)):
            while(mono_stack and temperatures[mono_stack[-1]]<temperatures[i]):
                diff=i-mono_stack[-1]
                output[mono_stack[-1]]=diff
                mono_stack.pop()
            mono_stack.append(i)
        return output 